crates/bevy_remote/src/lib.rs

```diff
@@ -0,0 +1,961 @@
+    components: Vec<String>,
+}
+
+/// The response to a `QUERY` request.
+#[derive(Serialize, Deserialize, Clone)]
+struct BrpQueryResponse {
+    /// All results of the query: the entities and the requested components.
+    rows: Vec<BrpQueryRow>,
+}
+
+/// One query match result: a single entity paired with the requested components.
+#[derive(Serialize, Deserialize, Clone)]
+pub struct BrpQueryRow {
+    /// The ID of the entity that matched.
+    pub entity: Entity,
+
+    /// The serialized values of the requested components.
+    #[serde(flatten)]
+    pub components: HashMap<String, Value>,
+}
+
+/// A message from the Bevy Remote Protocol server thread to the main world.
+///
+/// This is placed in the [`BrpMailbox`].
+#[derive(Clone)]
+pub struct BrpMessage {
+    /// The deserialized request from the client.
+    request: BrpRequest,
+
+    /// The channel on which the response is to be sent.
+    ///
+    /// The value sent here is serialized and sent back to the client.
+    sender: Arc<Mutex<Option<OneshotSender<AnyhowResult<Value>>>>>,
+}
+
+/// A resource that receives messages sent by Bevy Remote Protocol clients.
+///
+/// Every frame, the `process_remote_requests` system drains this mailbox, and
+/// processes the messages within.
+#[derive(Resource, Deref, DerefMut)]
+pub struct BrpMailbox(BroadcastReceiver<BrpMessage>);
+
+impl Default for RemotePlugin {
+    fn default() -> Self {
+        RemotePlugin { port: DEFAULT_PORT }
+    }
+}
+
+impl Default for RemoteVerbs {
+    fn default() -> Self {
+        let mut remote_verbs = RemoteVerbs(default());
+        remote_verbs.insert("GET".to_owned(), process_remote_get_request);
+        remote_verbs.insert("QUERY".to_owned(), process_remote_query_request);
+        remote_verbs.insert("SPAWN".to_owned(), process_remote_spawn_request);
+        remote_verbs.insert("INSERT".to_owned(), process_remote_insert_request);
+        remote_verbs.insert("REMOVE".to_owned(), process_remote_remove_request);
+        remote_verbs.insert("DESTROY".to_owned(), process_remote_destroy_request);
+        remote_verbs.insert("REPARENT".to_owned(), process_remote_reparent_request);
+        remote_verbs.insert("LIST".to_owned(), process_remote_list_request);
+        remote_verbs
+    }
+}
+
+impl Plugin for RemotePlugin {
+    fn build(&self, app: &mut App) {
+        app.insert_resource(RemotePort(self.port))
+            .init_resource::<RemoteVerbs>()
+            .add_systems(Startup, start_server)
+            .add_systems(Update, process_remote_requests);
+    }
+}
+
+/// A system that starts up the Bevy Remote Protocol server.
+fn start_server(mut commands: Commands, remote_port: Res<RemotePort>) {
+    // Create the channel and the mailbox.
+    let (request_sender, request_receiver) = broadcast::channel(CHANNEL_SIZE);
+    commands.insert_resource(BrpMailbox(request_receiver));
+
+    let port = remote_port.0;
+    thread::spawn(move || server_main(port, request_sender));
```

Comment:
This was mentioned once before, but is it possible to reuse the existing `IoTaskPool` here?

Reply:
Maybe, but let's make a decision as to whether to use Tokio or smol first.

Reply:
I now use the `IoTaskPool` and smol.